### PR TITLE
Add EncryptedAssertion decryption, shibboleth support

### PR DESCRIPTION
**Pull request by andrewstuart**
- Add hook to be able to choose authn request signing algorithm, for Shibboleth support
- Split files out for clarity
- Add decryption of encrypted assertions, supporting Shibboleth assertions
- Add appropriate testing for new integrations.

Completes #3, and an important step for coreos/dex#112. This would also be very useful for hashicorp/vault#370.

I'd imagine this is about 95% complete, and that @phoebesimon and @russellhaering will likely have some good suggestions for improvement, but I've gotten some good tests added, passing, and I've been able to authenticate fully using Shibboleth in my [toy application](/andrewstuart/saml2-toy-app).

Definitely let me know what I can do to improve things!

### CODE REVIEW

File: build_request.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	authnRequest.CreateAttr("ID", "_"+uuid.NewV4().String())
+	authnRequest.CreateAttr("Version", "2.0")
+	authnRequest.CreateAttr("ProtocolBinding", "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST")
+	authnRequest.CreateAttr("AssertionConsumerServiceURL", sp.AssertionConsumerServiceURL)
+	authnRequest.CreateAttr("AssertionConsumerServiceIndex", "0")
+	authnRequest.CreateAttr("AttributeConsumingServiceIndex", "0")
+	authnRequest.CreateAttr("IssueInstant", time.Now().UTC().Format(issueInstantFormat))
+
+	authnRequest.CreateAttr("Destination", "http://idp.astuart.co/idp/profile/SAML2/Redirect/SSO")
+	// authnRequest.CreateAttr("Destination", sp.IdentityProviderSSOURL)
```
Comment (phoebesimon) [inline build_request.go]: leftover?

Comment (andrewstuart) [inline build_request.go]: Ack, yes. Good catch. :-)

**Comment by andrewstuart** [comment]
I think this may be a tip-specific flake, by the way. Tests are still passing on 1.5 and 1.6. For whatever reason tip doesn't want to load the require test library.

**Comment by russellhaering** [comment]
Agreed regarding the test failure.

This looks good to me - awesome its working with Shibboleth!

**Pull request merged**